docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,331 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+callback is to allow for more complex and efficient thread handling. Utilizing a Task
+object allows for a producer to separate the "decoding" logic from the I/O, enabling a
+consumer to avoid transferring data between CPU cores (e.g. from one L1/L2 cache to another).
+
+This producer-provided structure has the following fields:
+
+.. c:member:: int (*ArrowArrayTask.get_data)(struct ArrowArrayTask*, struct ArrowDeviceArray*)
+
+  *Mandatory.* A callback to populate the provided ``ArrowDeviceArray`` with the available data.
+  The order of ArrowAsyncTasks provided by the producer enables a consumer to know the order of
+  the data to process.
+
+  If a non-zero value is returned from this, it should be followed only by the producer calling
+  the ``on_error`` callback of the ``ArrowAsyncDeviceStreamHandler``. Because calling this method
+  is likely to be separate from the current control flow, returning a non-zero value to signal
+  an error occuring allows the current thread to decide handle the case accordingly, while still
+  allowing all error logging and handling to be centralized in the
+  :c:member:`ArrowAsyncDeviceStreamHandler.on_error` callback.
+
+  Rather than having a separate release callback, any required cleanup should be performed as part
+  of the invocation of this callback, with ownership of the Array being passed to the consumer-provided
+  parameter calling this, and must be released separately.
+
+  It is only valid to call this method exactly once.
+
+.. c:member:: void* ArrowArrayTask.private_data
+
+  *Optional.* An opaque pointer to consumer-provided private data.
+
+  Consumers *MUST NOT* process this member. Lifetime of this member is handled by
+  the producer who created this object, and should be cleaned up if necessary during
+  the call to :c:member:`ArrowArrayTask.get_data`.
+
+The ArrowAsyncProducer structure
+''''''''''''''''''''''''''''''''
+
+This producer-provided and managed object has the following fields:
+
+.. c:member:: void (*ArrowAsyncProducer.request)(struct ArrowAsyncProducer*, uint64_t)
+
+  *Mandatory.* This function must be called by a consumer to start receiving calls to
+  :c:member:`ArrowAsyncDeviceStreamHandler.on_next_task`. It *MUST* be valid to call
+  this synchronously from within :c:member:`ArrowAsyncDeviceStreamHandler.on_next_task`
+  or :c:member:`ArrowAsyncDeviceStreamHandler.on_schema`. As a result, this function
+  *MUST NOT* synchronously call ``on_next_task`` or ``on_error`` to avoid recursive
+  and reentrant callbacks.
+
+  After ``cancel`` is called, additional calls to this function must be a NOP, but allowed.
+
+  While not cancelled, calling this function registers the given number of additional
+  arrays/batches to be produced by the producer. A producer should only call
+  the appropriate ``on_next_task`` callback up to a maximum of the total sum of calls to
+  this method before propagating back-pressure / waiting.
```

Comment:
Why does this talk about "propagating back-pressure"? If a consumer requests N arrays, it should receive N arrays.

Reply:
because of the disconnect between the consumer, the producer and whatever the *source* of the data is.

For instance, if the data is coming from a stream on the producer's side then just because the consumer requests N arrays, doesn't mean the producer won't receive >N arrays from the data source. Thus, back-pressure potentially needs to be handled by the producer to push back on the data source. This may be as simple as the producer *only* retrieving N arrays from its source, or it might be more complex. I didn't want to proscribe a sleep or any other particular way of handling the back-pressure.

Does that make more sense?

Reply:
Ah, I see.